### PR TITLE
Rebuild Permission System from main cloud as a module

### PR DESCRIPTION
**Pull request by TheMeinerLP**
Some improvements coming later as a extra rework.
On the rework must be do, recursive implementation of groups.

### CODE REVIEW

File: PermissionCore/src/main/java/eu/cloudnetsrvice/cloudpermission/model/CloudPlayer.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+import com.google.gson.reflect.TypeToken;
+
+import java.lang.reflect.Type;
+import java.sql.Timestamp;
+import java.util.Objects;
+
+public class CloudPlayer extends OfflinePlayer {
+
+    public static final Type TYPE = (new TypeToken<CloudPlayer>() {
```
Comment (GiantTreeLP) [inline PermissionCore/src/main/java/eu/cloudnetsrvice/cloudpermission/model/CloudPlayer.java]: Please use the TypeToken.get method here

File: PermissionCore/src/main/java/eu/cloudnetsrvice/cloudpermission/model/CloudPlayer.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        } else if (!(o instanceof CloudPlayer)) {
+            return false;
+        } else {
+            CloudPlayer that = (CloudPlayer) o;
+            return Objects.equals(this.playerConnection, that.playerConnection) && Objects.equals(this.proxy, that.proxy) && Objects.equals(
+                this.server,
+                that.server) && Objects.equals(this.loginTimeStamp, that.loginTimeStamp) && Objects.equals(this.playerExecutor,
+                that.playerExecutor);
+        }
+    }
```
Comment (GiantTreeLP) [inline PermissionCore/src/main/java/eu/cloudnetsrvice/cloudpermission/model/CloudPlayer.java]: Please recreate the equals method with IntelliJ

File: PermissionModule/src/main/java/eu/cloudnetservice/cloudpermission/module/config/ConfigPermissions.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.*;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+public class ConfigPermissions {
+
+    private static final Type COLLECTION_PERMISSION_GROUP_TYPE = TypeToken.getParameterized(Collection.class, PermissionGroup.class)
```
Comment (GiantTreeLP) [inline PermissionModule/src/main/java/eu/cloudnetservice/cloudpermission/module/config/ConfigPermissions.java]: We should move to concrete interfaces like List or Set in the future

File: PermissionPlugin/src/main/java/eu/cloudnetservice/cloudpermission/packet/out/PacketOutUpdatePlayer.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package eu.cloudnetservice.cloudpermission.packet.out;
+
+import de.dytanic.cloudnet.lib.network.protocol.packet.Packet;
+import de.dytanic.cloudnet.lib.utility.document.Document;
+import eu.cloudnetsrvice.cloudpermission.model.OfflinePlayer;
+
+public class PacketOutUpdatePlayer extends Packet {
+    public PacketOutUpdatePlayer(OfflinePlayer offlinePlayer) {
+        super(302, new Document("player", offlinePlayer));
```
Comment (GiantTreeLP) [inline PermissionPlugin/src/main/java/eu/cloudnetservice/cloudpermission/packet/out/PacketOutUpdatePlayer.java:9]: I don't like this arbitrary number

File: PermissionPlugin/src/main/java/eu/cloudnetservice/cloudpermission/packet/out/PacketOutUpdatePermissionGroup.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package eu.cloudnetservice.cloudpermission.packet.out;
+
+import de.dytanic.cloudnet.lib.network.protocol.packet.Packet;
+import de.dytanic.cloudnet.lib.utility.document.Document;
+import eu.cloudnetsrvice.cloudpermission.model.PermissionGroup;
+
+public class PacketOutUpdatePermissionGroup extends Packet {
+    public PacketOutUpdatePermissionGroup(PermissionGroup permissionGroup) {
+        super(201, new Document("permissionGroup", permissionGroup));
```
Comment (GiantTreeLP) [inline PermissionPlugin/src/main/java/eu/cloudnetservice/cloudpermission/packet/out/PacketOutUpdatePermissionGroup.java:9]: I don't like this arbitrary number

File: PermissionPlugin/src/main/java/eu/cloudnetservice/cloudpermission/listener/bungee/ProxiedPlayerListener.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        } else if (e.getSender() instanceof CloudPlayerCommandSender) {
+            e.setHasPermission(((CloudPlayerCommandSender) e.getSender()).getCloudPlayer()
+                                                                         .getPermissionEntity()
+                                                                         .hasPermission(PermissionAPI.getInstance().getPermissionPool(),
+                                                                             e.getPermission(),
+                                                                             CloudAPI.getInstance().getGroup()));
+        }
+    }
+
+}
```
Comment (GiantTreeLP) [inline PermissionPlugin/src/main/java/eu/cloudnetservice/cloudpermission/listener/bungee/ProxiedPlayerListener.java:38]: Please reformat this file

**Review by GiantTreeLP — CHANGES_REQUESTED** [review]

File: PermissionPlugin/src/main/java/eu/cloudnetservice/cloudpermission/packet/out/PacketOutUpdatePlayer.java (hunk 4 above)
Comment (TheMeinerLP) [reply PermissionPlugin/src/main/java/eu/cloudnetservice/cloudpermission/packet/out/PacketOutUpdatePlayer.java:9]: It only a copy from original code

File: PermissionPlugin/src/main/java/eu/cloudnetservice/cloudpermission/packet/out/PacketOutUpdatePermissionGroup.java (hunk 5 above)
Comment (TheMeinerLP) [reply PermissionPlugin/src/main/java/eu/cloudnetservice/cloudpermission/packet/out/PacketOutUpdatePermissionGroup.java:9]: It only a copy from original code

File: PermissionPlugin/src/main/java/eu/cloudnetservice/cloudpermission/listener/bungee/ProxiedPlayerListener.java (hunk 6 above)
Comment (TheMeinerLP) [reply PermissionPlugin/src/main/java/eu/cloudnetservice/cloudpermission/listener/bungee/ProxiedPlayerListener.java:38]: I have already do it.

**Review by GiantTreeLP — APPROVED** [review]

**Pull request merged**